File: backend/app/schemas/carrier.py
import json
from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator


class CarrierMetricsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    original_filename: str
    width: int
    height: int
    pixel_count: int
    raw_capacity_bytes: int
    max_payload_bytes: int
    shannon_entropy: float
    edge_density: float
    distortion_risk: float
    capacity_score: float
    entropy_score: float
    edge_score: float
    distortion_score: float
    overall_score: float
    explanation: list[str]
    created_at: datetime

    @field_validator("explanation", mode="before")
    @classmethod
    def _parse_explanation(cls, value: object) -> object:
        # The ORM stores explanation as a JSON-encoded string; decode it here
        # so API responses always expose a real list of strings.
        if isinstance(value, str):
            return json.loads(value)
        return value


class CarrierRankResponse(BaseModel):
    carriers: list[CarrierMetricsResponse]
    recommended: list[CarrierMetricsResponse]

File: frontend/src/types/carrier.ts
export interface CarrierMetrics {
  id: string;
  original_filename: string;
  width: number;
  height: number;
  pixel_count: number;
  raw_capacity_bytes: number;
  max_payload_bytes: number;
  shannon_entropy: number;
  edge_density: number;
  distortion_risk: number;
  capacity_score: number;
  entropy_score: number;
  edge_score: number;
  distortion_score: number;
  overall_score: number;
  explanation: string[];
  created_at: string;
}

export interface CarrierRankResponse {
  carriers: CarrierMetrics[];
  recommended: CarrierMetrics[];
}
